# Preprocessing and Exploritory Data Analysis

In [928]:
# Importing the necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import re 

In [929]:
# Load in the dataframe
df = pd.read_csv('../data/db_computers_2025_raw.csv')

/var/folders/_m/kfvhsj6515q3c7pdw8kq_qwc0000gn/T/ipykernel_98203/2794974568.py:2: DtypeWarning: Columns (6,7,8,9,13,14,17,18,19,20,21,22,24,25,26,27,28,29,30,32,36,37,40,41,42,46,49,50,51,53,59,60,62,63,67,70,73,75,76,77,78,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,98,100,103,104,105,126,127,128,129,130,132,133,134,135) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/db_computers_2025_raw.csv')


## Data inpection & summary statistics 

### Initial examination

In [930]:
df.head()


Unnamed: 0                            Título             Precio_Rango  \
0           0     Apple MacBook Air 13" 2024 M3  1.026,53 € – 2.287,17 €   
1           1     Apple MacBook Air 15" 2024 M3  1.270,42 € – 2.672,25 €   
2           2   Apple MacBook Pro 14" M4 (2024)  1.731,44 € – 4.932,25 €   
3           3  ASUS TUF Gaming A15 FA506 (2024)    647,00 € – 1.031,71 €   
4           4         GigaByte Aero 16 OLED BKF  1.031,31 € – 2.069,99 €   

        Ofertas     Tipo de producto              Serie  \
0  200 ofertas:  Portátil multimedia  Apple MacBook Air   
1  230 ofertas:  Portátil multimedia  Apple MacBook Air   
2  136 ofertas:  Portátil multimedia  Apple MacBook Pro   
3   30 ofertas:      Portátil gaming    ASUS TUF Gaming   
4   29 ofertas:      Portátil gaming      Gigabyte Aero   

  Pantalla_Tamaño de la pantalla Pantalla_Diagonal de la pantalla  \
0                  13,6 pulgadas                        34,544 cm   
1                  15,3 pulgadas                        38,862 cm   
2                  14,2 pulgadas                        36,068 cm   
3                  15,6 pulgadas                        39,624 cm   
4                    16 pulgadas                         40,64 cm   

  Pantalla_Tecnología de la pantalla Pantalla_Luminosidad  ...  \
0                             Retina            500 cd/m²  ...   
1                             Retina            500 cd/m²  ...   
2                             Retina          1.000 cd/m²  ...   
3                            Full HD                  NaN  ...   
4                               UHD+            600 cd/m²  ...   

  Características especiales Almacenamiento_Formato audiovisual compatible  \
0                        NaN                                           NaN   
1                        NaN                                           NaN   
2                        NaN                                           NaN   
3                        NaN                                           NaN   
4                        NaN                                           NaN   

  Otras características_Modo de sonido  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  Propiedades de la carcasa_Superficie del material  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  Disco duro_Capacidad disco duro (GB)  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  Propiedades de la carcasa_Bahía externa de 3,5" Procesador_Placa base  \
0                                             NaN                   NaN   
1                                             NaN                   NaN   
2                                             NaN                   NaN   
3                                             NaN                   NaN   
4                                             NaN                   NaN   

  Otras características_Certificado RoHS Sistema operativo_Software incluido  \
0                                    NaN                                 NaN   
1                                    NaN                                 NaN   
2                                    NaN                                 NaN   
3                                    NaN                                 NaN   
4                                    NaN                                 NaN   

  Procesador_Procesadores compatibles  
0                                 NaN  
1                                 NaN  
2                    

Based on the head of the dataset, we can see that the dataset has large amount of collumns. This is most likley due to the different types of PC parts that have their own unique properties that do not apply to other categories.

In [931]:
df.tail()

Unnamed: 0                                           Título  \
8059        4027              Ankermann Work Home (4260561745772)   
8060        4028             Captiva Advanced Gaming I75\nI75-363   
8061        4029  Vibox VIII-49 Gaming PC i9-11900F / RTX 4070 Ti   
8062        4030  Sedatech UCCG810I2I1HF - Watercooling Gaming-PC   
8063        4031              Ankermann Gaming PC (4260561745598)   

     Precio_Rango    Ofertas Tipo de producto Serie  \
8059     249,00 €  1 oferta:    PC de oficina   NaN   
8060   1.129,99 €  1 oferta:        PC gaming   NaN   
8061   2.249,95 €  1 oferta:        PC gaming   NaN   
8062   3.269,90 €  1 oferta:        PC gaming   NaN   
8063   2.543,00 €  1 oferta:        PC gaming   NaN   

     Pantalla_Tamaño de la pantalla Pantalla_Diagonal de la pantalla  \
8059                            NaN                              NaN   
8060                            NaN                              NaN   
8061                            NaN                              NaN   
8062                            NaN                              NaN   
8063                            NaN                              NaN   

     Pantalla_Tecnología de la pantalla Pantalla_Luminosidad  ...  \
8059                                NaN                  NaN  ...   
8060                                NaN                  NaN  ...   
8061                                NaN                  NaN  ...   
8062                                NaN                  NaN  ...   
8063                                NaN                  NaN  ...   

     Características especiales Almacenamiento_Formato audiovisual compatible  \
8059                        NaN                                           NaN   
8060                        NaN                                           NaN   
8061                        NaN                                           NaN   
8062                        NaN                                           NaN   
8063                        NaN                                           NaN   

     Otras características_Modo de sonido  \
8059                                  NaN   
8060                                  NaN   
8061                                  NaN   
8062                                  NaN   
8063                                  NaN   

     Propiedades de la carcasa_Superficie del material  \
8059                                               NaN   
8060                                               NaN   
8061                                               NaN   
8062                                               NaN   
8063                                               NaN   

     Disco duro_Capacidad disco duro (GB)  \
8059                               500 GB   
8060                                  NaN   
8061                                  NaN   
8062                                  NaN   
8063                                  NaN   

     Propiedades de la carcasa_Bahía externa de 3,5" Procesador_Placa base  \
8059                                             NaN                   NaN   
8060                                             NaN                   NaN   
8061                                             NaN                   NaN   
8062                                             NaN                   NaN   
8063                                             NaN                   NaN   

     Otras características_Certificado RoHS  \
8059                                    NaN   
8060                                    NaN   
8061                                    NaN   
8062                                    NaN   
8063                                    NaN   

     Sistema operativo_Software incluido Procesador_Procesadores compatibles  
8059                                 NaN                                 NaN  
8060                                 NaN                                 NaN  
8061                                 NaN     

In [932]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8064 entries, 0 to 8063
Columns: 136 entries, Unnamed: 0 to Procesador_Procesadores compatibles
dtypes: float64(10), int64(1), object(125)
memory usage: 8.4+ MB


There are 8063 items in this dataset that we will have to deal with. An important part to look at is the data types. We have 10 Columns with type Float, 1 column with an integer type, and 125 object columns. These datatypes are important to help us understand what type of database is suitable for this dataset.

### Summary Statistics

In [933]:
df.describe()

Unnamed: 0  Sonido_Número de altavoces  \
count  8064.000000                 2938.000000   
mean   2015.500000                    2.477195   
std    1164.010282                    1.111884   
min       0.000000                    2.000000   
25%    1007.750000                    2.000000   
50%    2015.500000                    2.000000   
75%    3023.250000                    2.000000   
max    4031.000000                    8.000000   

       Otras características_Fecha de lanzamiento  \
count                                 3781.000000   
mean                                  2022.613065   
std                                      1.511238   
min                                   2012.000000   
25%                                   2022.000000   
50%                                   2023.000000   
75%                                   2024.000000   
max                                   2025.000000   

       Disco duro_Número de discos duros (instalados)  \
count                                     5598.000000   
mean                                         0.988567   
std                                          0.238087   
min                                          0.000000   
25%                                          1.000000   
50%                                          1.000000   
75%                                          1.000000   
max                                          3.000000   

       Alimentación_Número de celdas  Procesador_Número de hilos de ejecución  \
count                    2679.000000                              3102.000000   
mean                        3.530795                                16.374275   
std                         1.037607                                 7.292888   
min                         2.000000                                 1.000000   
25%                         3.000000                                12.000000   
50%                         3.000000                                16.000000   
75%                         4.000000                                20.000000   
max                         8.000000                                32.000000   

       RAM_Número de ranuras para memoria RAM  \
count                             1043.000000   
mean                                 2.478428   
std                                  1.081277   
min                                  1.000000   
25%                                  2.000000   
50%                                  2.000000   
75%                                  2.000000   
max                                  8.000000   

       Propiedades de la carcasa_Bahía interna de 2,5"  \
count                                       358.000000   
mean                                          1.304469   
std                                           0.603010   
min                                           1.000000   
25%                                           1.000000   
50%                                           1.000000   
75%                                           2.000000   
max                                           6.000000   

       Propiedades de la carcasa_Bahía interna de 3,5"  \
count                                       138.000000   
mean                                          1.557971   
std                                           0.704709   
min                                           1.000000   
25%                                           1.000000   
50%                                           1.000000   
75%                                           2.000000   
max                                           5.000000   

       Propiedades de la carcasa_Bahía externa de 5,25"  \
count                                         49.000000   
mean                                           1.632653   
std                                            0.528121   
min                                            1.000000   
25%                                            1.000

Since not every row has values for all the columns, the .describe function does not help us as much as it could. We may have to split the dataset into multiple datasets to get a better understanding of the data.

In [934]:
df.nunique()


Unnamed: 0                                         4032
Título                                             8053
Precio_Rango                                       6706
Ofertas                                             110
Tipo de producto                                     17
                                                   ... 
Propiedades de la carcasa_Bahía externa de 3,5"       2
Procesador_Placa base                                13
Otras características_Certificado RoHS                1
Sistema operativo_Software incluido                   5
Procesador_Procesadores compatibles                   4
Length: 136, dtype: int64

There seems to be almost zero non-unqiue features. There are 8064 entries and 8053 of them are unique. All have their own values for properties that differeniate eachother. This is helpful to know that there are almost zero duplicate PC products. Let check now for any duplicates.

In [935]:
df.duplicated().sum()

0

Despite the number of entries (8064) and unqiue titles (8053) only slightly differing, there are no exact duplicate rows. So there is no further duplicate data handling needed.

In [936]:
# Check for columns that have no values
df.isnull().sum()

Unnamed: 0                                            0
Título                                                0
Precio_Rango                                        137
Ofertas                                             137
Tipo de producto                                    545
                                                   ... 
Propiedades de la carcasa_Bahía externa de 3,5"    8047
Procesador_Placa base                              8036
Otras características_Certificado RoHS             8057
Sistema operativo_Software incluido                8058
Procesador_Procesadores compatibles                8058
Length: 136, dtype: int64

Here we can see that there are no empty columns. We have a total of 8064 rows, and none of the columns have 8064 null values.

## Renaming Columns

Since the names of the columns are extremely inconsistent and repetative, lets perform some operations to make sure they are simple, consistent, and follow python best practices.

In [937]:
import re
from unidecode import unidecode

# common small words to drop after chopping off the repeated category
STOP_WORDS = {'de', 'del', 'la', 'el', 'los', 'las', 'y', 'a', 'en', 'para', 'por', 'con'}

def slugify(text: str) -> str:
    text = unidecode(text.strip())
    text = re.sub(r'[^0-9A-Za-z]+', '_', text)   # non-alnum → _
    text = re.sub(r'_+', '_', text)               # collapse __ → _
    return text.strip('_').lower()

def clean_col(col: str) -> str:
    col = col.strip()
    if "_" not in col:
        return slugify(col)
    cat_raw, attr_raw = col.split("_", 1)
    cat = slugify(cat_raw)
    attr = slugify(attr_raw)

    tokens = attr.split("_")
    if cat in tokens:
        # keep everything before the first appearance of the category
        idx = tokens.index(cat)
        tokens = tokens[:idx]
        # drop little filler words
        tokens = [t for t in tokens if t not in STOP_WORDS]
        attr = "_".join(tokens)

    return f"{cat}_{attr}" if attr else cat

# Apply toDataFrame:
df.columns = [clean_col(c) for c in df.columns]

df.head()

unnamed_0                            titulo             precio_rango  \
0          0     Apple MacBook Air 13" 2024 M3  1.026,53 € – 2.287,17 €   
1          1     Apple MacBook Air 15" 2024 M3  1.270,42 € – 2.672,25 €   
2          2   Apple MacBook Pro 14" M4 (2024)  1.731,44 € – 4.932,25 €   
3          3  ASUS TUF Gaming A15 FA506 (2024)    647,00 € – 1.031,71 €   
4          4         GigaByte Aero 16 OLED BKF  1.031,31 € – 2.069,99 €   

        ofertas     tipo_de_producto              serie pantalla_tamano  \
0  200 ofertas:  Portátil multimedia  Apple MacBook Air   13,6 pulgadas   
1  230 ofertas:  Portátil multimedia  Apple MacBook Air   15,3 pulgadas   
2  136 ofertas:  Portátil multimedia  Apple MacBook Pro   14,2 pulgadas   
3   30 ofertas:      Portátil gaming    ASUS TUF Gaming   15,6 pulgadas   
4   29 ofertas:      Portátil gaming      Gigabyte Aero     16 pulgadas   

  pantalla_diagonal pantalla_tecnologia pantalla_luminosidad  ...  \
0         34,544 cm              Retina            500 cd/m²  ...   
1         38,862 cm              Retina            500 cd/m²  ...   
2         36,068 cm              Retina          1.000 cd/m²  ...   
3         39,624 cm             Full HD                  NaN  ...   
4          40,64 cm                UHD+            600 cd/m²  ...   

  caracteristicas_especiales almacenamiento_formato_audiovisual_compatible  \
0                        NaN                                           NaN   
1                        NaN                                           NaN   
2                        NaN                                           NaN   
3                        NaN                                           NaN   
4                        NaN                                           NaN   

  otras_caracteristicas_modo_de_sonido  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  propiedades_de_la_carcasa_superficie_del_material  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  disco_duro_capacidad_disco_duro_gb  \
0                                NaN   
1                                NaN   
2                                NaN   
3                                NaN   
4                                NaN   

  propiedades_de_la_carcasa_bahia_externa_de_3_5 procesador_placa_base  \
0                                            NaN                   NaN   
1                                            NaN                   NaN   
2                                            NaN                   NaN   
3                                            NaN                   NaN   
4                                            NaN                   NaN   

  otras_caracteristicas_certificado_rohs sistema_operativo_software_incluido  \
0                                    NaN                                 NaN   
1                                    NaN                                 NaN   
2                                    NaN                                 NaN   
3                                    NaN                                 NaN   
4                                    NaN                                 NaN   

  procesador_procesadores_compatibles  
0                                 NaN  
1                                 NaN  
2                                 NaN  
3                                 NaN  
4                                 NaN  

[5 rows x 136 columns]

## Feature Cleaning

There are quite a few columns in this dataset that need some cleaning up. For example, the "Precio_Rango" is a string something like the following: "1.026,53 € – 2.287,17 €". This is not helpful for data processing. We need to turn this data into a framework that we can work with. In this section, we will go through every column and make sure the datatypes and values correctly match as intended.

#### "Unamed" Column

The "Unamed" Column simply represents the index of the item in the csv file. Since there are no relationships between the PC parts based on the index or a UUID, we can drop this column.

In [938]:
# Drop the "Unamed" column
df.drop(columns=['unnamed_0'], inplace=True)


#### "Precio_Rango"

The price column takes on two different forms: 

- **A range:** "1.026,53 € – 2.287,17 €"
- **A single number:** "479,00 €"

Despite the price of a product being a value, the column is a string, which is not helpful when querying for ranges and processing a value that is supposed to be a number.

Because of these factors, we are going to split this single string column into three different float columns:
- Precio_Max
- Precio_Min
- Precio_Mean

In [939]:
import numpy as np
import pandas as pd

def process_price_range(price_str):
    """
    Processes a price string (either a single price or a range)
    and returns min, max, and mean prices as floats.
    
    Handles formats like "1.026,53 € – 2.287,17 €" or "479,00 €".
    """
    if pd.isna(price_str):
        return np.nan, np.nan, np.nan

    # Standardize cleaning: remove currency, spaces, then fix separators
    cleaned_str = str(price_str).replace('€', '').strip()
    cleaned_str = cleaned_str.replace('.', '') # Remove thousands separator
    cleaned_str = cleaned_str.replace(',', '.') # Replace decimal separator

    min_price_str, max_price_str = None, None

    if '–' in cleaned_str: # It's a range
        parts = cleaned_str.split('–')
        min_price_str = parts[0].strip()
        max_price_str = parts[1].strip()
    else: # It's a single price
        min_price_str = cleaned_str
        max_price_str = cleaned_str
        
    try:
        min_price = float(min_price_str)
        max_price = float(max_price_str)
        mean_price = round((min_price + max_price) / 2, 2)  # Truncate to 2 decimal places
        return min_price, max_price, mean_price
    except (ValueError, TypeError):
        # In case of any conversion error after splitting/cleaning
        return np.nan, np.nan, np.nan

# replace the original column with the new columns
df['precio_min'], df['precio_max'], df['precio_mean'] = zip(*df['precio_rango'].apply(process_price_range))
df.drop(columns=['precio_rango'], inplace=True)

#### "Ofertas"

We want to clean the 'Ofertas' column in the dataset because it currently contains strings such as: 
- "200 ofertas:"
- "30 ofertas," 
- "1 oferta:" 

The goal is to extract only the numeric part, for example, 200 or 30, by removing any accompanying words or extra characters. Once cleaned, the values will be converted into float type to make them ready for numerical operations. This cleaning is necessary because the word "ofertas" is unnecessary for data analysis, and having clean numeric data allows us to easily perform tasks such as aggregating totals, calculating averages, visualizing the data, and comparing the number of offers between different products.

In [940]:
import numpy as np
import pandas as pd
import re # For regular expressions

def clean_ofertas(oferta_str):
    """
    Cleans the 'Ofertas' string to extract the number of offers.
    Converts "200 ofertas:", "30 ofertas,", "1 oferta:" into a float.
    """
    if pd.isna(oferta_str):
        return np.nan

    # Convert to string to be safe, in case of mixed types initially
    s = str(oferta_str)
    
    # Use regex to find the first sequence of digits
    match = re.search(r'\d+', s)
    
    if match:
        try:
            return float(match.group(0))
        except ValueError:
            # Should not happen if regex only matches digits, but good practice
            return np.nan
    else:
        # If no digits are found (e.g., an unexpected string like "No offers")
        return np.nan

# replace the original column
df['ofertas'] = df['ofertas'].apply(clean_ofertas)

#### "Pantalla_Tamaño de la pantalla"

For the 'Pantalla_Tamaño de la pantalla' column, we firstly want to clean the column name to be just "pantalla_tamano_pulgadas". Then, we aim to clean the data to make it more appropriate for numerical analysis. The column currently contains entries like:
- "13,6 pulgadas"
- "15,3 pulgadas," 

Which combine the screen size with its unit of measurement. The task is to isolate the numeric part of the string, such as 13.6 or 15.3, and remove any associated text or symbols. After cleaning, the values will be converted into floats.

In [941]:
import numpy as np
import pandas as pd
import re # For regular expressions

def clean_screen_size(size_str):
    """
    Cleans screen size strings like "13,6 pulgadas" or "15,3 pulgadas,"
    to extract the numeric value (e.g., 13.6) and convert it to a float.
    """
    if pd.isna(size_str):
        return np.nan

    s = str(size_str)
    
    # Replace comma decimal with period and remove non-numeric/non-period characters
    # This regex looks for a number that might have a comma as a decimal
    # and captures it. It's safer to first replace comma then extract.
    s = s.replace(',', '.')
    
    # Extract the first floating point number found
    # This regex matches an optional integer part, an optional decimal part,
    # or an integer part followed by a decimal part.
    match = re.search(r'\d+\.\d+|\d+', s) # Matches "13.6" or "13"
    
    if match:
        try:
            return float(match.group(0))
        except ValueError:
            return np.nan
    else:
        return np.nan

# 1. Rename the column first
# Make sure the old column name exists before trying to rename
original_col_name = 'pantalla_tamano'
new_col_name = 'pantalla_tamano_pulgadas'

if original_col_name in df.columns:
    df.rename(columns={original_col_name: new_col_name}, inplace=True)
    print(f"Column '{original_col_name}' renamed to '{new_col_name}'")

# 2. Clean the data in the newly renamed column
if new_col_name in df.columns: # Check if renaming was successful (or if it already existed)
    df[new_col_name] = df[new_col_name].apply(clean_screen_size)
    print(f"Cleaned data in '{new_col_name}' column.")
else:
    print(f"Column '{new_col_name}' not found after attempting rename.")

Column 'pantalla_tamano' renamed to 'pantalla_tamano_pulgadas'
Cleaned data in 'pantalla_tamano_pulgadas' column.


#### "Pantalla_Diagonal de la pantalla"

Firstly, we want to rename the "pantalla_diagonal" to simply "pantalla_diagonal_cm".

Secondly, the cleaning process involves extracting only the numeric values from strings that currently include both the size and the unit of measurement, such as "34,544 cm" or "40,64 cm." The objective is to remove the text and retain only the numeric component, such as 34.544 or 40.64, which will then be converted into floats.

In [942]:
import numpy as np
import pandas as pd
import re # For regular expressions

def clean_screen_diagonal(diagonal_str):
    """
    Cleans screen diagonal strings like "34,544 cm" or "40,64 cm,"
    to extract the numeric value (e.g., 34.544) and convert it to a float.
    """
    if pd.isna(diagonal_str):
        return np.nan

    s = str(diagonal_str)
    
    # Replace comma decimal with period.
    # It's important to do this before removing other non-numeric characters
    s = s.replace(',', '.') # Convert "34,544" to "34.544"
    
    # Extract the first floating point number found
    # This regex matches an optional integer part, an optional decimal part,
    # or an integer part followed by a decimal part.
    match = re.search(r'\d+\.\d+|\d+', s) # Matches "34.544" or "34"
    
    if match:
        try:
            return float(match.group(0))
        except ValueError:
            return np.nan
    else:
        return np.nan

# 1. Rename the column first
# Make sure the old column name exists before trying to rename
original_col_name_diag = "pantalla_diagonal"
new_col_name_diag = "pantalla_diagonal_cm"

if original_col_name_diag in df.columns:
    df.rename(columns={original_col_name_diag: new_col_name_diag}, inplace=True)
    print(f"Column '{original_col_name_diag}' renamed to '{new_col_name_diag}'")

    # 2. Clean the data in the newly renamed column
    if new_col_name_diag in df.columns: # Check if renaming was successful
        df[new_col_name_diag] = df[new_col_name_diag].apply(clean_screen_diagonal)
        print(f"Cleaned data in '{new_col_name_diag}' column.")
    else:
        print(f"Column '{new_col_name_diag}' not found after attempting rename.")
elif new_col_name_diag in df.columns: # If it was already named 'Pantalla_Diagonal'
    df[new_col_name_diag] = df[new_col_name_diag].apply(clean_screen_diagonal)
    print(f"Cleaned data in existing '{new_col_name_diag}' column.")
else:
    print(f"Original column '{original_col_name_diag}' not found. No action taken.")

Column 'pantalla_diagonal' renamed to 'pantalla_diagonal_cm'
Cleaned data in 'pantalla_diagonal_cm' column.


#### "Pantalla_Luminosidad"

We will clean the 'Pantalla_Luminosidad' column by removing the unit of measurement and any formatting symbols, such as "cd/m²" and thousand separators, to keep only the numeric brightness values as floats. This will allow us to perform accurate numerical operations like averaging, comparing, or visualizing brightness levels across devices. 

Secondly, after cleaning the data, we will rename the column to 'pantalla_luminosidad_cd_m2' to make the unit of measurement explicitly part of the column name. This ensures that, even though the values are now purely numeric, it remains clear that they represent brightness in candela per square meter, improving both clarity and documentation of the dataset.

In [943]:
def clean_luminosidad(luminosidad_str):
    """
    Cleans luminosity strings like "500 cd/m²" or "1.000 cd/m²"
    to extract the numeric value (e.g., 500 or 1000) and convert it to a float.
    """
    if pd.isna(luminosidad_str):
        return np.nan

    s = str(luminosidad_str)
    
    # Remove "cd/m²" and any variations of it (case-insensitive)
    s = re.sub(r'cd/m\^?2', '', s, flags=re.IGNORECASE).strip()
    
    # Remove thousands separators (dots or commas if they were used for that)
    s = s.replace('.', '') # Remove potential thousands separator
    s = s.replace(',', '.') # Convert potential decimal comma to period (if any)
                           # Although for luminosity, it's often whole numbers.

    # Extract the first sequence of digits (possibly with a decimal now)
    match = re.search(r'\d+\.?\d*|\d+', s) # Matches "500" or "500.5"
    
    if match:
        try:
            return float(match.group(0))
        except ValueError:
            return np.nan
    else:
        return np.nan


original_luminosidad_col = 'pantalla_luminosidad'
new_luminosidad_col = 'pantalla_luminosidad_cd_m2'

if original_luminosidad_col in df.columns:
    # 1. Clean the data in the original column
    df[original_luminosidad_col] = df[original_luminosidad_col].apply(clean_luminosidad)
    print(f"Cleaned data in '{original_luminosidad_col}'.")
    
    # 2. Rename the column
    df.rename(columns={original_luminosidad_col: new_luminosidad_col}, inplace=True)
    print(f"Column '{original_luminosidad_col}' renamed to '{new_luminosidad_col}'.")
else:
    print(f"Column '{original_luminosidad_col}' not found.")

Cleaned data in 'pantalla_luminosidad'.
Column 'pantalla_luminosidad' renamed to 'pantalla_luminosidad_cd_m2'.


#### "Grafica_Resolucion externa"

In the 'Gráfica_Resolución externa' column, we would standardize the format of the resolution values by making sure the resolution looks like: 1920x1080. This would make the data cleaner and more usable for numerical analysis and comparison of display resolutions. Removing any unnecessary characters like parentheses would also make the values more consistent.

In [944]:
original_column_name = 'grafica_resolucion_externa' 
new_column_name = 'grafica_resolucion_externa'

if original_column_name not in df.columns:
    print(f"Error: Column '{original_column_name}' not found in DataFrame.")
    print(f"Available columns are: {df.columns.tolist()}")
    # Verify the exact cleaned name of 'Gráfica_Resolución externa'
else:
    def standardize_external_resolution(resolution_str):
        """
        Standardizes an external resolution string to "WIDTHxHEIGHT" format.
        Example: "hasta 1920x1080 @ 60Hz" -> "1920x1080"
        Example: "4K (3840 x 2160)" -> "3840x2160"
        Returns the standardized string, or np.nan if not parsable.
        """
        if pd.isna(resolution_str):
            return np.nan

        s = str(resolution_str)
        
        # Regex to find patterns like "1920x1080" or "3840 x 2160"
        # It looks for two numbers (3 or more digits each) separated by 'x' (case-insensitive),
        # with optional spaces around the 'x'.
        # We take the first match found in the string.
        match = re.search(r'(\d{3,})\s*[xX]\s*(\d{3,})', s)
        
        if match:
            try:
                width = int(match.group(1))
                height = int(match.group(2))
                return f"{width}x{height}"
            except ValueError:
                # This case should be rare if the regex matches digits
                return np.nan
        else:
            # If the primary WxH pattern is not found, return NaN.
            # You could add more sophisticated parsing here if there are other common formats.
            return np.nan

    # Apply the function to the original column to create the new standardized column
    df[new_column_name] = df[original_column_name].apply(standardize_external_resolution)

    print(f"Created new standardized column: '{new_column_name}'")
    print("\nSample of original and new standardized columns:")
    
    # Display some rows where the original column is not NaN to see the effect
    sample_indices = df[original_column_name].notna()
    if sample_indices.any():
        # Create a temporary DataFrame for display, showing only relevant columns and non-NaN originals
        display_df = df.loc[sample_indices, [original_column_name, new_column_name]].copy()
        # For cleaner display, show unique transformations or just the head
        # This gets a bit complex if we want to show *every* unique transformation,
        # so let's show a sample of rows that were actually transformed.
        transformed_rows = display_df[display_df[new_column_name].notna()].head(10)
        if not transformed_rows.empty:
            print(transformed_rows)
        else:
            print(f"No rows were successfully transformed from '{original_column_name}'. Check patterns.")
            print(df[[original_column_name, new_column_name]].head())
    else:
        print(f"The column '{original_column_name}' seems to contain all NaN values.")
        print(df[[new_column_name]].head())


    print(f"\nValue counts for '{new_column_name}':")
    print(df[new_column_name].value_counts(dropna=False).head(10)) # Show top 10 results
    
    print(f"\nData type of '{new_column_name}': {df[new_column_name].dtype}")


    # just drop the original if the new one is sufficient:
    df.drop(columns=[original_column_name], inplace=True)
    print(f"\nOriginal column '{original_column_name}' dropped, use '{new_column_name}'.")


Created new standardized column: 'grafica_resolucion_externa'

Sample of original and new standardized columns:
   grafica_resolucion_externa grafica_resolucion_externa
0                   6016x3384                  6016x3384
1                   6016x3384                  6016x3384
2                   8192x4608                  8192x4608
3                   4096x2304                  4096x2304
9                   6016x3384                  6016x3384
23                  6016x3384                  6016x3384
25                  4096x2304                  4096x2304
31                  6016x3384                  6016x3384
35                  8192x4608                  8192x4608
39                  8192x4608                  8192x4608

Value counts for 'grafica_resolucion_externa':
grafica_resolucion_externa
NaN          7656
6016x3384     169
8192x4608     162
4096x2304      65
5120x2880       6
2560x1600       5
3840x2160       1
Name: count, dtype: int64

Data type of 'grafica_resolucion_

#### "Cámara_Resolución de la cámara"

In [945]:
# Check the unique values in the column
df['camara_resolucion'].unique()

array(['1920 x 1080\xa0píxeles', nan, '720 x 480\xa0píxeles',
       '1280 x 720\xa0píxeles', '640 x 480\xa0píxeles',
       '2560 x 1440\xa0píxeles', '2592 x 1944\xa0píxeles',
       '480 x 640\xa0píxeles'], dtype=object)

In [946]:
original_column_name = 'camara_resolucion' # Adjust if your cleaned name is different
new_column_name = 'camara_resolucion'

if original_column_name not in df.columns:
    print(f"Error: Column '{original_column_name}' not found in DataFrame.")
    print(f"Available columns are: {df.columns.tolist()}")
    # Verify the exact cleaned name of the camera resolution column.
else:
    def standardize_camera_resolution(resolution_str):
        """
        Standardizes a camera resolution string to "WIDTHxHEIGHT" format.
        Handles numeric patterns and common names like HD, Full HD, VGA.
        """
        if pd.isna(resolution_str):
            return np.nan

        s = str(resolution_str)
        s_lower = s.lower() # For case-insensitive keyword matching

        # Priority 1: Explicit WxH pattern (e.g., "1280x720", "1920 x 1080")
        # Regex looks for two numbers separated by 'x' (case-insensitive), with optional spaces.
        match = re.search(r'(\d+)\s*[xX]\s*(\d+)', s)
        if match:
            try:
                width = int(match.group(1))
                height = int(match.group(2))
                return f"{width}x{height}"
            except ValueError:
                # This case is unlikely if the regex matches digits but included for robustness
                pass 

        # If no known pattern or keyword is matched, return NaN
        return np.nan

    # Apply the function to the original column to create the new standardized column
    df[new_column_name] = df[original_column_name].apply(standardize_camera_resolution)

    print(f"Created new standardized column: '{new_column_name}'")
    print("\nSample of original and new standardized columns:")
    
    sample_indices = df[original_column_name].notna()
    if sample_indices.any():
        display_df = df.loc[sample_indices, [original_column_name, new_column_name]].copy()
        # Show a sample of rows that were actually transformed or had original values
        transformed_sample = display_df.drop_duplicates().head(10)
        if not transformed_sample.empty:
            print(transformed_sample)
        else:
            print(f"Could not generate a useful sample from '{original_column_name}'.")
            print(df[[original_column_name, new_column_name]].head())
    else:
        print(f"The column '{original_column_name}' seems to contain all NaN values or was empty.")
        print(df[[new_column_name]].head())

    print(f"\nValue counts for '{new_column_name}' (top 10):")
    print(df[new_column_name].value_counts(dropna=False).head(10))
    
    print(f"\nData type of '{new_column_name}': {df[new_column_name].dtype}")

    # After verifying, drop it:
    df.drop(columns=[original_column_name], inplace=True)
    print(f"\nOriginal column '{original_column_name}' dropped, use '{new_column_name}'.")

Created new standardized column: 'camara_resolucion'

Sample of original and new standardized columns:
     camara_resolucion camara_resolucion
0            1920x1080         1920x1080
11             720x480           720x480
15            1280x720          1280x720
150            640x480           640x480
652          2560x1440         2560x1440
1819         2592x1944         2592x1944
2413           480x640           480x640

Value counts for 'camara_resolucion' (top 10):
camara_resolucion
NaN          5270
1280x720     1453
1920x1080    1296
720x480        16
640x480        10
2560x1440      10
2592x1944       7
480x640         2
Name: count, dtype: int64

Data type of 'camara_resolucion': object

Original column 'camara_resolucion' dropped, use 'camara_resolucion'.


#### "Teclado_Número de teclas"

In [947]:
original_column_name = 'teclado_numero_de_teclas' 
new_column_name = 'teclado_numero_de_teclas_horas'

if original_column_name not in df.columns:
    print(f"Error: Column '{original_column_name}' not found in DataFrame.")
    print(f"Available columns are: {df.columns.tolist()}")
    # Verify the exact cleaned name of the 'teclado_numero' column.
else:
    def extract_number_of_keys(value_str):
        """
        Extracts the number of keys from a string like "79 teclas".
        Returns the number as a float, or np.nan if not parsable.
        """
        if pd.isna(value_str):
            return np.nan

        s = str(value_str)
        
        # Regex to find the first sequence of digits in the string.
        match = re.search(r'(\d+)', s)
        
        if match:
            try:
                return float(match.group(1)) # Using float for consistency with NaNs
            except ValueError:
                # This should not happen if regex only matches digits
                return np.nan
        else:
            # If no digits are found
            return np.nan

    # Apply the function to the original column to create the new numerical column
    df[new_column_name] = df[original_column_name].apply(extract_number_of_keys)

    print(f"Created new numerical column: '{new_column_name}'")
    print("\nSample of original and new columns:")
    
    sample_indices = df[original_column_name].notna()
    if sample_indices.any():
        display_df = df.loc[sample_indices, [original_column_name, new_column_name]].copy()
        # Show a sample of rows that were actually transformed or had original values
        transformed_sample = display_df.drop_duplicates().head(10)
        if not transformed_sample.empty:
            print(transformed_sample)
        else:
            print(f"Could not generate a useful sample from '{original_column_name}'.")
            print(df[[original_column_name, new_column_name]].head())
    else:
        print(f"The column '{original_column_name}' seems to contain all NaN values or was empty.")
        print(df[[new_column_name]].head())

    print(f"\nValue counts for '{new_column_name}' (top 10):")
    print(df[new_column_name].value_counts(dropna=False).head(10))
    
    print(f"\nData type of '{new_column_name}': {df[new_column_name].dtype}")

    # just drop the original if the new one is sufficient:
    df.drop(columns=[original_column_name], inplace=True)
    print(f"\nOriginal column '{original_column_name}' dropped, use '{new_column_name}'.")


Created new numerical column: 'teclado_numero_de_teclas_horas'

Sample of original and new columns:
     teclado_numero_de_teclas  teclado_numero_de_teclas_horas
0                   79 teclas                            79.0
20                  98 teclas                            98.0
278                 80 teclas                            80.0
686                 84 teclas                            84.0
873                 66 teclas                            66.0
1012                65 teclas                            65.0
1495                99 teclas                            99.0
1821               102 teclas                           102.0
2527               107 teclas                           107.0
3177                90 teclas                            90.0

Value counts for 'teclado_numero_de_teclas_horas' (top 10):
teclado_numero_de_teclas_horas
NaN      7772
79.0      233
65.0       13
98.0        8
66.0        8
99.0        8
80.0        5
107.0       5
84.0        4


#### "Alimentación_Autonomía de la batería"

In [948]:
df['alimentacion_autonomia_de_la_bateria'].unique()

array(['15\xa0h', '13\xa0h', nan, '8,4\xa0h', '5,3\xa0h', '5,5\xa0h',
       '23\xa0h', '8\xa0h', '7,5\xa0h', '14\xa0h', '7,25\xa0h',
       '4,5\xa0h', '10\xa0h', '27\xa0h', '7,6\xa0h', '6,75\xa0h',
       '10,25\xa0h', '12\xa0h', '9,7\xa0h', '7\xa0h', '16\xa0h',
       '20\xa0h', '6,25\xa0h', '17\xa0h', '12,5\xa0h', '8,25\xa0h',
       '9\xa0h', '29\xa0h', '18\xa0h', '13,4\xa0h', '29,5\xa0h', '4\xa0h',
       '19\xa0h', '24,5\xa0h', '6\xa0h', '23,5\xa0h', '15,5\xa0h',
       '15,3\xa0h', '11,5\xa0h', '4,75\xa0h', '10,5\xa0h', '13,5\xa0h',
       '18,5\xa0h', '21,4\xa0h', '22\xa0h', '11\xa0h', '7,4\xa0h',
       '11,75\xa0h', '8,5\xa0h', '15,25\xa0h', '13,6\xa0h', '8,1\xa0h',
       '5\xa0h', '9,75\xa0h', '13,25\xa0h', '17,5\xa0h', '9,5\xa0h',
       '26\xa0h', '11,1\xa0h', '23,25\xa0h', '6,5\xa0h', '14,5\xa0h',
       '12,8\xa0h', '20,5\xa0h', '11,25\xa0h', '15,4\xa0h', '9,25\xa0h',
       '9,4\xa0h', '8,75\xa0h'], dtype=object)

In [949]:
original_column_name = 'alimentacion_autonomia_de_la_bateria' # Adjust if your cleaned name is different
new_column_name = 'alimentacion_autonomia_horas'

if original_column_name not in df.columns:
    print(f"Error: Column '{original_column_name}' not found in DataFrame.")
    print(f"Available columns are: {df.columns.tolist()}")
    # Verify the exact cleaned name of the battery autonomy column.
else:
    def extract_battery_autonomy_hours(value_str):
        """
        Extracts the battery autonomy in hours from a string like "15\\xa0h" or "8,4\\xa0h".
        Returns the number of hours as a float, or np.nan if not parsable.
        """
        if pd.isna(value_str):
            return np.nan

        s = str(value_str)
        
        # Replace comma with period for decimal conversion
        s = s.replace(',', '.')
        
        # Use regex to find a number (integer or decimal)
        # This will also effectively strip "h" and "\xa0" if they follow the number.
        # The pattern looks for digits, optionally followed by a period and more digits.
        match = re.search(r'(\d+\.?\d*|\.\d+)', s)
        
        if match:
            try:
                # The first group of the match should be our number
                return float(match.group(1))
            except ValueError:
                # Should not happen if regex matches correctly
                return np.nan
        else:
            # If no number is found
            return np.nan

    # Apply the function to the original column to create the new numerical column
    df[new_column_name] = df[original_column_name].apply(extract_battery_autonomy_hours)

    print(f"Created new numerical column: '{new_column_name}'")
    print("\nSample of original and new columns:")
    
    sample_indices = df[original_column_name].notna()
    if sample_indices.any():
        display_df = df.loc[sample_indices, [original_column_name, new_column_name]].copy()
        transformed_sample = display_df.drop_duplicates().head(10) # Show unique transformations
        if not transformed_sample.empty:
            print(transformed_sample)
        else:
            print(f"Could not generate a useful sample from '{original_column_name}'.")
            print(df[[original_column_name, new_column_name]].head())
    else:
        print(f"The column '{original_column_name}' seems to contain all NaN values or was empty.")
        print(df[[new_column_name]].head())

    print(f"\nValue counts for '{new_column_name}' (top 10):")
    print(df[new_column_name].value_counts(dropna=False).head(10))
    
    print(f"\nData type of '{new_column_name}': {df[new_column_name].dtype}")

    # After verifying, you might want to replace the original column or drop it:
    # Example:
    # df.drop(columns=[original_column_name], inplace=True)
    # print(f"\nOriginal column '{original_column_name}' dropped, use '{new_column_name}'.")

Created new numerical column: 'alimentacion_autonomia_horas'

Sample of original and new columns:
   alimentacion_autonomia_de_la_bateria  alimentacion_autonomia_horas
0                                  15 h                         15.00
2                                  13 h                         13.00
10                                8,4 h                          8.40
12                                5,3 h                          5.30
19                                5,5 h                          5.50
20                                 23 h                         23.00
33                                  8 h                          8.00
48                                7,5 h                          7.50
54                                 14 h                         14.00
59                               7,25 h                          7.25

Value counts for 'alimentacion_autonomia_horas' (top 10):
alimentacion_autonomia_horas
NaN     6333
15.0     211
10.0     170
7.5  

#### "Alimentación_Vatios-hora"

In [950]:
df['alimentacion_vatios_hora'].unique()

array(['52,6\xa0Wh', '66,5\xa0Wh', '70\xa0Wh', '48\xa0Wh', '88\xa0Wh',
       '75\xa0Wh', '90\xa0Wh', '38\xa0Wh', '50\xa0Wh', '60\xa0Wh',
       '73\xa0Wh', '54\xa0Wh', '42\xa0Wh', '72\xa0Wh', '99\xa0Wh',
       '86\xa0Wh', '76\xa0Wh', '45\xa0Wh', '65\xa0Wh', '83\xa0Wh',
       '53,5\xa0Wh', '35\xa0Wh', '63\xa0Wh', '53,8\xa0Wh', '68\xa0Wh',
       '47\xa0Wh', '100\xa0Wh', '80\xa0Wh', '53\xa0Wh', nan, '61\xa0Wh',
       '41\xa0Wh', '95\xa0Wh', '57\xa0Wh', '67\xa0Wh', '78\xa0Wh',
       '52,4\xa0Wh', '51,3\xa0Wh', '77\xa0Wh', '48,5\xa0Wh', '52\xa0Wh',
       '48,8\xa0Wh', '73,6\xa0Wh', '56\xa0Wh', '71\xa0Wh', '40\xa0Wh',
       '52,5\xa0Wh', '55\xa0Wh', '99,9\xa0Wh', '43\xa0Wh', '36\xa0Wh',
       '46\xa0Wh', '53,3\xa0Wh', '45,8\xa0Wh', '82\xa0Wh', '39\xa0Wh',
       '64\xa0Wh', '37\xa0Wh', '66\xa0Wh', '74\xa0Wh', '94\xa0Wh',
       '63,5\xa0Wh', '48,96\xa0Wh', '59\xa0Wh', '84\xa0Wh', '51\xa0Wh',
       '58\xa0Wh', '42,75\xa0Wh', '58,2\xa0Wh', '47,4\xa0Wh',
       '51,5\xa0Wh', '92\xa0Wh

In [951]:
import pandas as pd
import numpy as np
import re

# Assume 'df' is your DataFrame.
# The column name 'alimentacion_vatios_hora' should have been cleaned by your clean_col function.
original_column_name = 'alimentacion_vatios_hora' # Adjust if your cleaned name is different
new_column_name = 'alimentacion_vatios_hora'

if original_column_name not in df.columns:
    print(f"Error: Column '{original_column_name}' not found in DataFrame.")
    print(f"Available columns are: {df.columns.tolist()}")
    # Verify the exact cleaned name of the Watt-hours column.
else:
    def extract_watt_hours(value_str):
        """
        Extracts the Watt-hour value from a string like "45 Wh" or "35,5\\xa0Wh".
        Returns the value as a float, or np.nan if not parsable.
        """
        if pd.isna(value_str):
            return np.nan

        s = str(value_str)
        
        # Replace comma with period for decimal conversion
        s = s.replace(',', '.')

        match = re.search(r'(\d+\.?\d*|\.\d+)', s)
        
        if match:
            try:
                # The first group of the match should be our number
                return float(match.group(1))
            except ValueError:
                # Should not happen if regex matches correctly
                return np.nan
        else:
            # If no number is found
            return np.nan

    # Apply the function to the original column to create the new numerical column
    df[new_column_name] = df[original_column_name].apply(extract_watt_hours)

    print(f"Created new numerical column: '{new_column_name}'")
    print("\nSample of original and new columns:")
    
    sample_indices = df[original_column_name].notna()
    if sample_indices.any():
        display_df = df.loc[sample_indices, [original_column_name, new_column_name]].copy()
        # Show unique transformations for a better overview
        transformed_sample = display_df.drop_duplicates().head(10) 
        if not transformed_sample.empty:
            print(transformed_sample)
        else:
            print(f"Could not generate a useful sample from '{original_column_name}'.")
            print(df[[original_column_name, new_column_name]].head())
    else:
        print(f"The column '{original_column_name}' seems to contain all NaN values or was empty.")
        print(df[[new_column_name]].head())

    print(f"\nValue counts for '{new_column_name}' (top 10):")
    print(df[new_column_name].value_counts(dropna=False).head(10))
    
    print(f"\nData type of '{new_column_name}': {df[new_column_name].dtype}")

    df.drop(columns=[original_column_name], inplace=True)
    print(f"\nOriginal column '{original_column_name}' dropped, use '{new_column_name}'.")


Created new numerical column: 'alimentacion_vatios_hora'

Sample of original and new columns:
    alimentacion_vatios_hora  alimentacion_vatios_hora
0                       52.6                      52.6
1                       66.5                      66.5
2                       70.0                      70.0
3                       48.0                      48.0
4                       88.0                      88.0
5                       75.0                      75.0
6                       90.0                      90.0
10                      38.0                      38.0
11                      50.0                      50.0
12                      60.0                      60.0

Value counts for 'alimentacion_vatios_hora' (top 10):
alimentacion_vatios_hora
NaN     4398
41.0     315
90.0     243
42.0     203
70.0     171
57.0     157
54.0     141
56.0     130
38.0     120
83.0     120
Name: count, dtype: int64

Data type of 'alimentacion_vatios_hora': float64

Original colum

## Dataset Splitting

In [952]:
df['tipo_de_producto'].unique()

array(['Portátil multimedia', 'Portátil gaming', 'Portátil profesional',
       nan, 'Ultrabook', 'Portátil convertible', 'Netbook', 'Portátil 3D',
       'Chromebook', 'Mini PC', 'PC multimedia', 'PC gaming',
       'PC de oficina', 'Workstation', 'Barebone', 'PC completo',
       'Thin Client', 'Kit ampliación PC'], dtype=object)

Because different "Tipo de producto" has extremely different properties, it will be beneficial to split the dataset into different csv files. This will help us will object and schema managemnet when storing and querying for items in the database.

### Split by 'Tipo De Producto'

In [953]:
unique_product_types = df['Tipo_de_Producto'].unique()
dfs_by_product_type = {}

for product_type in unique_product_types:
    if pd.isna(product_type): 
        subset_df = df[df['Tipo_de_Producto'].isna()].copy() 
        subset_df.dropna(axis=1, how='all', inplace=True) # Clean it too
        dfs_by_product_type['Desconocido'] = subset_df 
        print(f"Processed 'Desconocido': original columns {df[df['Tipo_de_Producto'].isna()].shape[1]}, cleaned columns {subset_df.shape[1]}")
    else:
        subset_df = df[df['Tipo_de_Producto'] == product_type].copy()
        original_cols = subset_df.shape[1]
        # Drop columns where ALL values are NaN
        subset_df.dropna(axis=1, how='all', inplace=True) 
        cleaned_cols = subset_df.shape[1]
        dfs_by_product_type[product_type] = subset_df
        print(f"Processed '{product_type}': original columns {original_cols}, cleaned columns {cleaned_cols}")


KeyError: 'Tipo_de_Producto'

In [503]:
import os

output_data_path = '../data/split_by_type/' 
if not os.path.exists(output_data_path):
    os.makedirs(output_data_path)

# The dfs_by_product_type dictionary now holds the cleaned DataFrames
for product_type, df_cleaned_subset in dfs_by_product_type.items():
    filename_safe_product_type = "".join(c if c.isalnum() else "_" for c in str(product_type))
    file_path = os.path.join(output_data_path, f"{filename_safe_product_type}.csv")
    df_cleaned_subset.to_csv(file_path, index=False)
    print(f"Saved cleaned data to: {file_path}")


Saved cleaned data to: ../data/split_by_type/Portátil_multimedia.csv
Saved cleaned data to: ../data/split_by_type/Portátil_gaming.csv
Saved cleaned data to: ../data/split_by_type/Portátil_profesional.csv
Saved cleaned data to: ../data/split_by_type/Desconocido.csv
Saved cleaned data to: ../data/split_by_type/Ultrabook.csv
Saved cleaned data to: ../data/split_by_type/Portátil_convertible.csv
Saved cleaned data to: ../data/split_by_type/Netbook.csv
Saved cleaned data to: ../data/split_by_type/Portátil_3D.csv
Saved cleaned data to: ../data/split_by_type/Chromebook.csv
Saved cleaned data to: ../data/split_by_type/Mini_PC.csv
Saved cleaned data to: ../data/split_by_type/PC_multimedia.csv
Saved cleaned data to: ../data/split_by_type/PC_gaming.csv
Saved cleaned data to: ../data/split_by_type/PC_de_oficina.csv
Saved cleaned data to: ../data/split_by_type/Workstation.csv
Saved cleaned data to: ../data/split_by_type/Barebone.csv
Saved cleaned data to: ../data/split_by_type/PC_completo.csv
Saved 

### Split by Category

In [504]:
# Example mapping:
category_map = {
    'Portátil multimedia': 'Laptop', 'Portátil gaming': 'Laptop', 
    'Portátil profesional': 'Laptop', 'Ultrabook': 'Laptop',
    'Portátil convertible': 'Laptop', 'Netbook': 'Laptop',
    'Portátil 3D': 'Laptop', 'Chromebook': 'Laptop',

    'Mini PC': 'Desktop_PC', 'PC multimedia': 'Desktop_PC', 
    'PC gaming': 'Desktop_PC', 'PC de oficina': 'Desktop_PC',
    'Workstation': 'Desktop_PC', 'Barebone': 'Desktop_PC', 
    'PC completo': 'Desktop_PC',

    'Kit ampliación PC': 'Component_Kit',
    'Thin Client': 'Specialized_PC', 
    # Add others as needed
}

In [505]:
df['Categoria_General'] = df['Tipo_de_Producto'].map(category_map)
# Handle any types not in your map, perhaps by assigning them to 'Other'
df['Categoria_General'].fillna('Otro', inplace=True) 

/var/folders/_m/kfvhsj6515q3c7pdw8kq_qwc0000gn/T/ipykernel_98203/1913369206.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Categoria_General'].fillna('Otro', inplace=True)


In [ ]:
dfs_by_general_category = {}
for general_category_name in df['Categoria_General'].unique():
    subset_df = df[df['Categoria_General'] == general_category_name].copy()
    # IMPORTANT: Drop all-NaN columns again for these new broader DFs
    subset_df.dropna(axis=1, how='all', inplace=True)
    dfs_by_general_category[general_category_name] = subset_df
    
    # Save these new broader category DataFrames to new CSVs
    # (You'll want a new output folder or different filenames)
    output_path_general = '../data/split_by_general_category/'
    if not os.path.exists(output_path_general):
        os.makedirs(output_path_general)
    
    filename_safe_general_category = "".join(c if c.isalnum() else "_" for c in str(general_category_name))
    file_path = os.path.join(output_path_general, f"{filename_safe_general_category}.csv")
    subset_df.to_csv(file_path, index=False)
    print(f"Saved cleaned data to: {file_path} (Rows: {len(subset_df)}, Columns: {subset_df.shape[1]})")

Saved cleaned data to: ../data/split_by_general_category/Laptop.csv (Rows: 3889, Columns: 108)
Saved cleaned data to: ../data/split_by_general_category/Otro.csv (Rows: 545, Columns: 125)
Saved cleaned data to: ../data/split_by_general_category/Desktop_PC.csv (Rows: 3602, Columns: 72)
Saved cleaned data to: ../data/split_by_general_category/Specialized_PC.csv (Rows: 25, Columns: 49)
Saved cleaned data to: ../data/split_by_general_category/Component_Kit.csv (Rows: 3, Columns: 33)


## Visualization

## Transformation

## Feature Engineering

## Validation & Consistency Checks